### ISSUE TITLE
Standalone igtl Client

### ISSUE THREAD

**Issue by christiana** [IGSIO/OpenIGTLinkIF#6]
I have been considering how to extract a full standalone igtl Client from the OpenIGTLinkIF repo, as discussed in Boston (http://wiki.na-mic.org/Wiki/index.php/2016_Winter_Project_Week/Projects/TrackedUltrasoundStandardization).

For testing and possibly for use in applications, we would like the GUI from OpenIGTLink as Slicer-independent code. There should also be possible to reuse more basic parts of the library, either the full Client (no GUI), or only the converter classes. Suggested dependencies below:
![igtlclient_uml](https://cloud.githubusercontent.com/assets/436646/12401708/826b9464-be28-11e5-8d38-50bc69b61c35.png)

All the Slicer-independent code could ultimately be included in the OpenIGTLink repo itself, every module as a conditional build.

The Client itself can be separated from Slicer like shown in the diagram:
![igtlclient_uml](https://cloud.githubusercontent.com/assets/436646/12401999/617b102a-be2a-11e5-9527-72192a4d9fcd.png)

The main idea is to remove the inheritance from MRMLNode ( currently Connector, Converterts and Devices all are MRMLNodes), and instead use aggregation, i.e. Slicer will listen to the Client and create MRMLNodes wrapping the igtl devices/connectors as needed. The Converters could be an implementation detail hidden from external users. The Manager would contain much of the code currently present in vtkSlicerOpenIGTLinkIFLogic. A data layer duplicating some of MRML's features would have to be implemented (i.e. the Nodes), but this can be made as simple as possible, with no persistence (for now).

I am very interested in comments to this proposal, in particular: 
- Do this (first) design outlined look sound, what can be improved?
- Who is interested in the resulting code, and at what level?
- Can the result eventually be merged back into Slicer? This determines whether or not we should refactor the Slicer binding layer or simply extract the code of interest.

**Comment by adamrankin** [comment]
Thank you very much for putting this effort into the IF layer. Is this something we would like to meet about virtually and discuss?

**Comment by drouin-simon** [comment]
I fully agree with the overall architecture that you propose in the first diagram. I would suggest we focus on the lower levels first (the client and the converters) and then, once its working, we can start talking about the interface and see if we can all agree on a single one for all our apps. 

About the way you separate Slicer code and standalone client code, this is more or less the structure I had in mind, but I find the Node class on the non-mrml side complicates things. I think most applications that want nothing to do with Slicer or MRML won't care that device and connection derive from the same class. Specialized nodes on the MRML side can have Connection and Device aggregates. But maybe I didn't understand what you were trying to do... we could a have a quick google hangout (say we limit it to 30 min) to discuss this.

**Comment by christiana** [comment]
@drouin-simon: The node actually was a way to ease the connection to Slicer, I see that now. The adaption should be on the other (Slicer) side.

**Comment by christiana** [comment]
I suggest we have a Google Hangout tomorrow January 20. at 20:00UTC (21:00 Norway, 15:00 Boston/Montreal). I'll post the link here.

**Comment by adamrankin** [comment]
Is there any chance it could be moved + or - one hour?

**Comment by christiana** [comment]
+1 hour is OK for me, i.e January 20. at 21:00UTC (22:00 Norway, 16:00 Boston/Montreal). Is that ok for all?

**Comment by adamrankin** [comment]
That would be great, thanks!

**Comment by drouin-simon** [comment]
Works for me.

**Comment by lassoan** [comment]
I'm available tomorrow any time before 15:45 (or after 5pm). I like Christian's proposal and agree with Simon's comments, so I'm quite sure that the discussions will go in the right directions even without me - no need to reschedule the meeting just because of my availability.

**Comment by christiana** [comment]
Ok, we'll stick with the current time. Great that you follow the discussion, Andras.

**Comment by christiana** [comment]
New structure. All work is done within each device, except for creation. The converters are used privately by devices.

![igtlclient_uml](https://cloud.githubusercontent.com/assets/436646/12461354/bf6d7378-bfb8-11e5-9ac0-e81d82a2ba73.png)

One missing piece is the querying mechanism from Slicer, i.e. for GET_, STT, and getting replies. I cannot find any example of usage for this. In the above design all messages related to one device, also queries, should be handled by that device.

**Comment by christiana** [comment]
https://hangouts.google.com/call/etb6mvdshlhiiyuix6bcwu3b3aa

**Comment by adamrankin** [comment]
https://hangouts.google.com/call/h3pryazesc4jecb5ccmeupddpua

**Comment by christiana** [comment]
What are the opinions on Qt dependency? Are there some kind of policy that says Qt should only be in the GUI? 

There is some kind of separation in OpenIGTLinkIF: The Logic is QT-independent, while the module and gui are Qt-dependent. This causes some strange tricks to appear, like having a QTimer in the qSlicerOpenIGTLinkIFModule that calls into vtkSlicerOpenIGTLinkIFLogic, thus breaking encapsulation.

IF we chose to use Qt also in the backend, i.e. in the client, that would simplify the code a lot.

In CustusX we use Qt on all levels, btw.

**Comment by lassoan** [comment]
Qt is a huge dependency, much more difficult to accommodate than VTK:
1. VTK is often used with .Net GUI, more and more on mobile devices, etc. This would be much more difficult with Qt.
2. Recent free Qt versions are LGPLv3, which prevent using of the toolkit in locked-down devices (where the user cannot replace a library), such as in most medical devices. VTK is BSD license.
3. It's nice to be able to create a small VTK client without a GUI or with a single VTK render window. Qt code base is 10x or so larger, takes more time to build, etc.

So, yes, I would avoid using Qt in the logic.

**Comment by christiana** [comment]
Thanks Andras. Good arguments, in particular the licensing issue. I'll keep the logic Qt-free.

**Comment by drouin-simon** [comment]
I agree with Andras, though I like to delay that kind of design decision at least until you have a working prototype. Before then, I say: whatever works. Take the shortest path to this first working prototype. If it means keeping a dependency on Qt, then keep it. But avoid adding more dependencies on Qt. We will figure out the best way to remove the dependency on Qt once it is working.

**Comment by christiana** [comment]
I've just started writing, so I'll keep away from Qt from the beginning. This simply means that I have to spend some time getting more familiar with the vtk event mechanisms.

As a starting point, I have created a new repo OpenIGTLinkIO (split from OpenIGTLinkIF), which lies between OpenIGTLink and OpenIGTLinkIF. This saves a lot of configure time, as opposed to working within Slicer. The work in progress can be observed on github.com/christiana/OpenIGTLinkIO.git. I'll notify you when it is reviewable.

**Comment by drouin-simon** [comment]
You can configure OpenIGTLinkIF separate from Slicer. That's what I'm doing. Just need to tell it where your Slicer build is. I just disabled OpenIGTLinkIF in the Slicer build and build our version of it elsewhere. I'm just not sure how to tell Slicer where our OpenIGTLinkIF is. I'm still trying to figure out the ideal development setup.

**Comment by christiana** [comment]
The first step is now available on https://github.com/christiana/OpenIGTLinkIO. The working code now consist of the folders Converters, Logic, Devices, just check the root CMakeLists.txt to see what's in use. The other files are not in use and will be upgraded or removed in the next step.

The structure is now pretty much like in the last diagram in this thread, with vtkIGTLIOLogic as the manager of a set of Connections, each with a list of Devices. igtl message conversion is done by the Devices themselves. Each of the new classes have a short description in-code.

Main changes from the old OpenIGTLinkIF code:
- Independent of Slicer, only depends on igtl and vtk.
- removal of MRML gave a simpler data structure - much of the complexity in OpenIGTLinkIF lies in the interaction with MRML and a lack of an interface layer between the two.
- Only support for ImageDevice yet.
- Query mechanism (GET_, STT_, ...) was in a MRML node, now moved into the Device superclass. I could not find any users of this code, does anyone know a project where it is used?
- Creation of Devices are now in a Factory object.

The code compiles, but have not been tested at all, thus you should only look at the structure and not the details. Next steps are adding a standalone gui, then reconnecting to Slicer. Slicer reconnection would be the hardest step, but should give a slimmer OpenIGTLinkIF.

Comments are welcome. 

Mostly anything can be changed at this point, including repo organization. It might be easiest to migrate the end result into optional folders of OpenIGTLink itself, instead of having a separate repo.

**Comment by lassoan** [comment]
Thanks for the developments. Query mechanism is used for example by OpenIGTLinkRemote module (in SlicerIGT extension) for querying and retrieving images from BrainLab and MedTronic navigation systems.

**Comment by christiana** [comment]
Thanks, Andras. OpenIGTLinkRemote looks small and nicely written. I'll take it into account during the next steps.

**Comment by drouin-simon** [comment]
Christian, I have just created a pull request that adds a converter class for the TRANSFORM message. I pushed my changes to IGSIO/OpenIGTLinkIF, but you seem to have been working on christiana/OpenIGTLinkIO, which is a different project. Where should we merge everything? Perhaps you could create IGSIO/OpenIGTLinkIO where we can all put our changes. Let me know what you want to do. I don't have a preference as to how we should proceed, as long as I know where to commit my work. Also, I would like to start playing with your latest code on Linux.

**Comment by christiana** [comment]
Thanks Simon. Yes, I have switched to christiana/OpenIGTLinkIO in order to be completely removed from the Slicer code. The only reason I kept it on my own account was that I did not feel the work is ready to share yet. I am currently working on a GUI independent on Slicer, that can be built as a standalone (test) client or be included into another project. I will move the repo to IGSIO as soon as possible, so that we can work on the same repo.

**Comment by drouin-simon** [comment]
Cool. Let me know if I can help with the GUI.

**Comment by christiana** [comment]
I have now pushed what I have to https://github.com/IGSIO/OpenIGTLinkIO. The README contains some info. I have added issues for some deficits that I know about.

What is done so far: An example app containing GUI for the Connectors, similar to what you can find under "Connectors" in the Slicer OpenIGTLinkIF GUI. The GUI has been somewhat tested, but the Connection has not. I am trying to get used to the vtk event system, so that is not working fully yet.

Simon: It would be great if you could compile it (Mac first, then Linux), there are probably some issues there. Qt4-support would also be welcome. Both of these have been added as https://github.com/IGSIO/OpenIGTLinkIO/issues.

As to the actual igtl communication, I'll take a round or two getting it into a somewhat working state before bothering you with it.

Missing: GUI for all connectors, maybe some gui from SlicerIGT, debugging, add all devices.

**Comment by drouin-simon** [comment]
Will do. I was already trying to compile the one on your account. I will just switch my git Origin to IGSIO. 

b.t.w.: I see you require vtk 6.1. Is this the minimum or the one you are using? Do you build vtk SHARED libs?

**Comment by christiana** [comment]
I use SHARED libs. Right now I use VTK 6.3, but we should support whatever Slicer is using as well.

You will probably spend some time with CTK if this is your first time building it. 
I use:

CTK_QT_VERSION:STRING=5
CTK_ENABLE_DICOM:BOOL=ON
CTK_LIB_DICOM/Widgets:BOOL=ON
CTK_ENABLE_PluginFramework:BOOL=ON
BUILD_TESTING:BOOL=OFF
CTK_BUILD_SHARED_LIBS:BOOL=ON
VTK_DIR:PATH=path/to/vtk

Although DICOM and the plugin fw should be unneeded here. I use a build from 2015-10-19, simply because thats what CustusX uses atm.

The dependency on CTK is weak, only a few macros, but it made the port a bit simpler.

**Comment by drouin-simon** [comment]
CTK is a pain to build. Grrr...

**Comment by drouin-simon** [comment]
Lesson so far: don't rely on VTK being built by CTK superbuild. It is trying to build VTK with Qt5 but somehow ends up building it with Qt4 and then CTK build fails.

**Comment by christiana** [comment]
I had the same experience. Add VTK_DIR:PATH=path/to/your/already/built/vtk before building CTK, this will override the superbuild VTK.
